File: src/markdown.py
"""
Markdown generation — Step 2 of the digest pipeline.

Reads all data/*.json files (written by collection.py) and writes a
{CITY_KEY.upper()}.md for each at the repo root.

Run: python src/markdown.py   (no env vars needed)
"""

import json
from datetime import date
from pathlib import Path

from common import CATEGORY_EMOJI, TOP_PICK_THRESHOLD, fmt_date


def write_markdown(
    events: list[dict], monday: date, sunday: date, city_name: str, city_key: str
) -> Path:
    top_picks = [e for e in events if e.get("score", 0) >= TOP_PICK_THRESHOLD]
    remaining = [e for e in events if e.get("score", 0) < TOP_PICK_THRESHOLD]

    lines: list[str] = []

    lines.append(f"# {city_name} — This Week's Events")
    lines.append(f"**{fmt_date(monday)} – {fmt_date(sunday)}**  ")
    lines.append(f"*{len(top_picks)} top picks · {len(events)} events total*")
    lines.append("")

    if top_picks:
        lines.append("## ⭐ Top Picks")
        lines.append("")
        for e in top_picks:
            cat    = e.get("category", "Community / Other")
            emoji  = CATEGORY_EMOJI.get(cat, "📌")
            cost   = e.get("cost", "See link")
            cost_s = "Free" if cost.lower() == "free" else cost
            tags   = e.get("tags", [])
            link   = e.get("link", "")

            title = e.get("title", "Untitled")
            lines.append(f"### {emoji} [{title}]({link})" if link else f"### {emoji} {title}")
            lines.append(f"📆 {e.get('datetime', '—')}  ")
            lines.append(f"📍 {e.get('location', '—')}  ")
            lines.append(f"💰 {cost_s}  ")
            if tags:
                lines.append("`" + "` `".join(tags[:6]) + "`")
            if desc := e.get("description", ""):
                lines.append("")
                lines.append(desc)
            lines.append("")

    if remaining:
        by_cat: dict[str, list[dict]] = {}
        for e in remaining:
            by_cat.setdefault(e.get("category", "Community / Other"), []).append(e)

        lines.append("## 📋 All Events")
        lines.append("")
        for cat, cat_events in by_cat.items():
            emoji = CATEGORY_EMOJI.get(cat, "📌")
            lines.append(f"### {emoji} {cat}")
            lines.append("")
            for e in cat_events:
                cost   = e.get("cost", "See link")
                cost_s = "Free" if cost.lower() == "free" else cost
                tags   = e.get("tags", [])
                link   = e.get("link", "")

                title = e.get("title", "Untitled")
                lines.append(f"#### [{title}]({link})" if link else f"#### {title}")
                lines.append(f"📆 {e.get('datetime', '—')}")
                lines.append(f"📍 {e.get('location', '—')}")
                lines.append(f"💰 {cost_s}")
                if tags:
                    lines.append("`" + "` `".join(tags[:6]) + "`")
                if desc := e.get("description", ""):
                    lines.append("")
                    lines.append(desc)
                lines.append("")

    out_path = Path(__file__).parent.parent / f"{city_key.upper()}.md"
    out_path.write_text("\n".join(lines), encoding="utf-8")
    print(f"→ Written {out_path.name} ({len(events)} events)")
    return out_path


def main() -> None:
    data_dir = Path(__file__).parent.parent / "data"
    json_files = [f for f in sorted(data_dir.glob("*.json")) if f.name != "index.json"]

    if not json_files:
        raise SystemExit("✗ No city JSON files found in data/ — run collection.py first.")

    print("Markdown generation")
    print("=" * 50)

    for json_path in json_files:
        payload = json.loads(json_path.read_text(encoding="utf-8"))
        events  = payload["events"]
        monday  = date.fromisoformat(payload["week_start"])
        sunday  = date.fromisoformat(payload["week_end"])
        write_markdown(
            events, monday, sunday,
            city_name=payload["city"],
            city_key=payload["city_key"],
        )

    print("✓ Markdown complete.")


if __name__ == "__main__":
    main()

File: src/markdown.ts
import { readdirSync, readFileSync, writeFileSync } from "node:fs";
import { join } from "node:path";

import {
	CATEGORY_EMOJI,
	DATA_ROOT,
	fmtDate,
	PROJECT_ROOT,
	TOP_PICK_THRESHOLD,
} from "./common.ts";

type Event = Record<string, unknown>;

function writeMarkdown(
	events: Event[],
	monday: Date,
	sunday: Date,
	cityName: string,
	cityKey: string,
): void {
	const topPicks = events.filter(
		(e) => ((e.score as number) ?? 0) >= TOP_PICK_THRESHOLD,
	);
	const remaining = events.filter(
		(e) => ((e.score as number) ?? 0) < TOP_PICK_THRESHOLD,
	);

	const lines: string[] = [];

	lines.push(`# ${cityName} — This Week's Events`);
	lines.push(`**${fmtDate(monday)} – ${fmtDate(sunday)}**  `);
	lines.push(`*${topPicks.length} top picks · ${events.length} events total*`);
	lines.push("");

	if (topPicks.length > 0) {
		lines.push("## ⭐ Top Picks");
		lines.push("");
		for (const e of topPicks) {
			const cat = (e.category as string) ?? "Community / Other";
			const emoji = CATEGORY_EMOJI[cat] ?? "📌";
			const cost = (e.cost as string) ?? "See link";
			const costS = cost.toLowerCase() === "free" ? "Free" : cost;
			const tags = (e.tags as string[]) ?? [];
			const link = (e.link as string) ?? "";
			const title = (e.title as string) ?? "Untitled";

			lines.push(
				link ? `### ${emoji} [${title}](${link})` : `### ${emoji} ${title}`,
			);
			lines.push(`📆 ${(e.datetime as string) ?? "—"}  `);
			lines.push(`📍 ${(e.location as string) ?? "—"}  `);
			lines.push(`💰 ${costS}  `);
			if (tags.length > 0) {
				lines.push(`\`${tags.slice(0, 6).join("` `")}\``);
			}
			const desc = (e.description as string) ?? "";
			if (desc) {
				lines.push("");
				lines.push(desc);
			}
			lines.push("");
		}
	}

	if (remaining.length > 0) {
		const byCat: Record<string, Event[]> = {};
		for (const e of remaining) {
			const cat = (e.category as string) ?? "Community / Other";
			if (!byCat[cat]) byCat[cat] = [];
			byCat[cat].push(e);
		}

		lines.push("## 📋 All Events");
		lines.push("");
		for (const [cat, catEvents] of Object.entries(byCat)) {
			const emoji = CATEGORY_EMOJI[cat] ?? "📌";
			lines.push(`### ${emoji} ${cat}`);
			lines.push("");
			for (const e of catEvents) {
				const cost = (e.cost as string) ?? "See link";
				const costS = cost.toLowerCase() === "free" ? "Free" : cost;
				const tags = (e.tags as string[]) ?? [];
				const link = (e.link as string) ?? "";
				const title = (e.title as string) ?? "Untitled";

				lines.push(link ? `#### [${title}](${link})` : `#### ${title}`);
				lines.push(`📆 ${(e.datetime as string) ?? "—"}`);
				lines.push(`📍 ${(e.location as string) ?? "—"}`);
				lines.push(`💰 ${costS}`);
				if (tags.length > 0) {
					lines.push(`\`${tags.slice(0, 6).join("` `")}\``);
				}
				const desc = (e.description as string) ?? "";
				if (desc) {
					lines.push("");
					lines.push(desc);
				}
				lines.push("");
			}
		}
	}

	const outPath = join(PROJECT_ROOT, `${cityKey.toUpperCase()}.md`);
	writeFileSync(outPath, lines.join("\n"), "utf-8");
	console.log(
		`→ Written ${cityKey.toUpperCase()}.md (${events.length} events)`,
	);
}

function main(): void {
	const jsonFiles = readdirSync(DATA_ROOT)
		.filter(
			(f) =>
				f.endsWith(".json") && f !== "index.json" && !f.includes("_raw.json"),
		)
		.sort()
		.map((f) => join(DATA_ROOT, f));

	if (jsonFiles.length === 0) {
		throw new Error(
			"✗ No city JSON files found in data/ — run collection.ts first.",
		);
	}

	console.log("Markdown generation");
	console.log("=".repeat(50));

	for (const jsonPath of jsonFiles) {
		const payload = JSON.parse(readFileSync(jsonPath, "utf-8"));
		writeMarkdown(
			payload.events,
			new Date(payload.week_start),
			new Date(payload.week_end),
			payload.city,
			payload.city_key,
		);
	}

	console.log("✓ Markdown complete.");
}

main();
